ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -47,16 +53,90 @@ var (
 	// headless svc will be of the form: {kuberay-cluster-name}-headless-worker-svc
 	headlessServiceSuffix = "headless-worker-svc"
 
-	// map of pod slices to workers in the slice
-	sliceToWorkers map[slice][]worker
-
 	// Flag arguments.
-	BindAddr   string
-	CACert     string
-	ServerCert string
-	ServerKey  string
+	BindAddr       string
+	CACert         string
+	KubeConfigPath string
+	ServerCert     string
+	ServerKey      string
 )
 
+func NewTPUWebhookServer(podLister listersv1.PodLister) *TPUWebhookServer {
+	return &TPUWebhookServer{
+		podLister: podLister,
+	}
+}
+
+func (t *TPUWebhookServer) Mutate(w http.ResponseWriter, r *http.Request) {
+	admissionReview := &admissionv1.AdmissionReview{}
+	if err := json.NewDecoder(r.Body).Decode(admissionReview); err != nil {
+		http.Error(w, "Error decoding request body", http.StatusBadRequest)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if admissionReview.Request.Kind.Kind != "Pod" {
+		http.Error(w, "Invalid Kind", http.StatusBadRequest)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	klog.V(1).Info("Received review for Pod creation")
+	response, err := t.mutatePod(admissionReview)
+	if err != nil {
+		klog.Errorf("Failed to mutate pod: %s", err)
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+	admissionReview.Response = response
+	responseBytes, err := json.Marshal(admissionReview)
+	if err != nil {
+		klog.Errorf("Failed to encode response: %s", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	fmt.Fprint(w, string(responseBytes))
+}
+
+func (t *TPUWebhookServer) Validate(w http.ResponseWriter, r *http.Request) {
+	admissionReview := &admissionv1.AdmissionReview{}
+	if err := json.NewDecoder(r.Body).Decode(admissionReview); err != nil {
+		http.Error(w, "Error decoding request body", http.StatusBadRequest)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if admissionReview.Request.Kind.Kind != "RayCluster" {
+		http.Error(w, "Invalid Kind", http.StatusBadRequest)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	klog.V(0).Info("Received review for RayCluster")
+	response, err := validateRayCluster(admissionReview)
+	if err != nil {
+		klog.Errorf("Failed to validate ray cluster: %s", err)
+		w.WriteHeader(http.StatusForbidden)
```

Comment:
StatusInternalServerError instead?

Reply:
I think StatusForbidden might make more sense here, since it indicates that the server understands the request but can't authorize it (since the validation fails) whereas StatusInternalServerError  seems like more of a catch-all for other errors

Reply:
I think requests that fail validation shouldn't return 403 or 500, it should return a valid AdmissionResponse but set `allowed = false` and let apiserver reject it properly.

Reply:
Ah I see we do this already, so an error returned from `validateRayCluster` is not failed validation, but rather an unexpected error, in which case it should be `StatusInternalServerError`?

Reply:
fwiw it doesn't really matter as long as validation fails :)

Reply:
Synced offline, StatusForbidden is fine if validation fails, but as a follow-up change we should include the status code and error in AdmissionResponse, e.g.: https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/webhook/admission/response.go#L64